docs/tutorials/go-built-in.md

```diff
@@ -103,579 +101,527 @@ func NewKVStoreApplication() *KVStoreApplication {
  return abcitypes.ResponseBeginBlock{}
 }
 
-func (KVStoreApplication) EndBlock(req abcitypes.RequestEndBlock) abcitypes.ResponseEndBlock {
+func (app *KVStoreApplication) EndBlock(req abcitypes.RequestEndBlock) abcitypes.ResponseEndBlock {
  return abcitypes.ResponseEndBlock{}
 }
 
-func (KVStoreApplication) ListSnapshots(abcitypes.RequestListSnapshots) abcitypes.ResponseListSnapshots {
+func (app *KVStoreApplication) ListSnapshots(abcitypes.RequestListSnapshots) abcitypes.ResponseListSnapshots {
  return abcitypes.ResponseListSnapshots{}
 }
 
-func (KVStoreApplication) OfferSnapshot(abcitypes.RequestOfferSnapshot) abcitypes.ResponseOfferSnapshot {
+func (app *KVStoreApplication) OfferSnapshot(abcitypes.RequestOfferSnapshot) abcitypes.ResponseOfferSnapshot {
  return abcitypes.ResponseOfferSnapshot{}
 }
 
-func (KVStoreApplication) LoadSnapshotChunk(abcitypes.RequestLoadSnapshotChunk) abcitypes.ResponseLoadSnapshotChunk {
+func (app *KVStoreApplication) LoadSnapshotChunk(abcitypes.RequestLoadSnapshotChunk) abcitypes.ResponseLoadSnapshotChunk {
  return abcitypes.ResponseLoadSnapshotChunk{}
 }
 
-func (KVStoreApplication) ApplySnapshotChunk(abcitypes.RequestApplySnapshotChunk) abcitypes.ResponseApplySnapshotChunk {
+func (app *KVStoreApplication) ApplySnapshotChunk(abcitypes.RequestApplySnapshotChunk) abcitypes.ResponseApplySnapshotChunk {
  return abcitypes.ResponseApplySnapshotChunk{}
 }
 ```
 
-Now I will go through each method explaining when it's called and adding
-required business logic.
+### 1.3.1 Add a persistent data store
 
-### 1.3.1 CheckTx
+Our application will need to write its state out to persistent storage so that it
+can stop and start without losing all of its data.
+
+For this tutorial, we will use [badger](https://github.com/dgraph-io/badger). 
+Badger is a fast embedded key-value store. 
+
+First, add badger as a dependency of your go module using the `go get` command:
+
+`go get github.com/dgraph-io/badger`
+
+Next, let's update the application and its constructor to receive a handle to the 
+database. 
 
-When a new transaction is added to the Tendermint Core, it will ask the
-application to check it (validate the format, signatures, etc.).
+Update the application struct as follows: 
 
 ```go
-import "bytes"
-
-func (app *KVStoreApplication) isValid(tx []byte) (code uint32) {
- // check format
- parts := bytes.Split(tx, []byte("="))
- if len(parts) != 2 {
-  return 1
- }
-
- key, value := parts[0], parts[1]
-
- // check if the same key=value already exists
- err := app.db.View(func(txn *badger.Txn) error {
-  item, err := txn.Get(key)
-  if err != nil && err != badger.ErrKeyNotFound {
-   return err
-  }
-  if err == nil {
-   return item.Value(func(val []byte) error {
-    if bytes.Equal(val, value) {
-     code = 2
-    }
-    return nil
-   })
-  }
-  return nil
- })
- if err != nil {
-  panic(err)
- }
-
- return code
+type KVStoreApplication struct {
+	db           *badger.DB
+	currentBatch *badger.Txn
```

Comment:
Would `currentBlock` be more accurate in this case? It is, as you note, a batch of transactions, but from the app's perspective that's sort of an implementation detail.

Maybe also worth a comment here:

```suggestion
	currentBatch *badger.Txn  // accumulates the transactions in the current block
```